[BUG]EmptyHashedRelation$ cannot be cast to org.apache.spark.sql.rapids.execution.SerializeConcatHostBuffersDeserializeBatch

**Describe the bug**
Caused by: java.lang.ClassCastException: org.apache.spark.sql.execution.joins.EmptyHashedRelation$ cannot be cast to org.apache.spark.sql.rapids.execution.SerializeConcatHostBuffersDeserializeBatch
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$2(GpuSubqueryBroadcastExec.scala:196)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withExecutionId$1(SQLExecution.scala:139)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.withExecutionId(SQLExecution.scala:137)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$1(GpuSubqueryBroadcastExec.scala:182)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

--- a/integration_tests/src/main/python/dpp_test.py
+++ b/integration_tests/src/main/python/dpp_test.py
@@ -300,3 +300,30 @@ def test_dpp_from_swizzled_hash_keys(spark_tmp_table_factory, aqe_enabled):
         conf=dict(_dpp_conf + [('spark.sql.adaptive.enabled', aqe_enabled),
                                ("spark.rapids.sql.castStringToTimestamp.enabled", "true"),
                                ("spark.rapids.sql.hasExtendedYearValues", "false")]))
+
+# Test handling DPP subquery that could broadcast EmptyRelation rather than a GPU serialized batch
+@pytest.mark.parametrize('aqe_enabled', [
+    'false',
+    pytest.param('true', marks=pytest.mark.skipif(is_before_spark_320(),
+                                                  reason='Only in Spark 3.2.0+ AQE and DPP can be both enabled'))
+], ids=idfn)
+@pytest.mark.skipif(is_databricks_runtime(), reason='DPP is not supported on Databricks runtime')
+def test_dpp_empty_relation(spark_tmp_table_factory, aqe_enabled):
+    dim_table = spark_tmp_table_factory.get()
+    fact_table = spark_tmp_table_factory.get()
+    def setup_tables(spark):
+        spark.sql("CREATE TABLE {}(id string) PARTITIONED BY (dt date, hr string, mins string) STORED AS PARQUET".format(dim_table))
+        spark.sql("INSERT INTO {}(id,dt,hr,mins) values ('somevalue', date('2022-01-01'), '11', '59')".format(dim_table))
+        spark.sql("CREATE TABLE {}(id string)".format(fact_table) +
+                  " PARTITIONED BY (dt date, hr string, mins string) STORED AS PARQUET")
+        spark.sql("INSERT INTO {}(id,dt,hr,mins)".format(fact_table) +
+                  " SELECT 'somevalue', to_date('2022-01-01'), '11', '59'")
+    with_cpu_session(setup_tables, conf={
+        "hive.exec.dynamic.partition" : "true",
+        "hive.exec.dynamic.partition.mode" : "nonstrict"
+    })
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark : spark.sql("SELECT COUNT(*) AS cnt FROM {} f".format(fact_table) +
+                                 " LEFT JOIN (SELECT * from {}) tmp".format(dim_table) +
+                                 " ON f.hr = tmp.hr AND f.dt = tmp.dt WHERE tmp.mins > 60"),
+        conf=dict(_dpp_conf + [('spark.sql.adaptive.enabled', aqe_enabled)]))

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
@@ -207,42 +207,52 @@ case class GpuSubqueryBroadcastExec(
       // This will run in another thread. Set the execution id so that we can connect these jobs
       // with the correct execution.
       SQLExecution.withExecutionId(sparkSession, executionId) {
-        val beforeCollect = System.nanoTime()
-
-        val serBatch = child.executeBroadcast[SerializeConcatHostBuffersDeserializeBatch]()
-
-        // Creates projection to extract target field from Row, as what Spark does.
-        // Note that unlike Spark, the GPU broadcast data has not applied the key expressions from
-        // the HashedRelation, so that is applied here if necessary to ensure the proper values
-        // are being extracted. The CPU already has the key projections applied in the broadcast
-        // data and thus does not have similar logic here.
-        val broadcastModeProject = modeKeys.map { keyExprs =>
-          val keyExpr = keyExprs(index)
-          UnsafeProjection.create(keyExpr)
+        val broadcastBatch = child.executeBroadcast[Any]()
+        val result: Array[InternalRow] = broadcastBatch.value match {
+          case b: SerializeConcatHostBuffersDeserializeBatch => projectSerializedBatchToRows(b)
+          case b if SparkShimImpl.isEmptyRelation(b) => Array.empty
+          case b => throw new IllegalStateException(s"Unexpected broadcast type: ${b.getClass}")
         }
-
-        // Use the single output of the broadcast mode projection if it exists
-        val rowProjectIndex = if (broadcastModeProject.isDefined) 0 else index
-        val rowProject = UnsafeProjection.create(
-          BoundReference(rowProjectIndex, buildKeys(index).dataType, buildKeys(index).nullable))
-
-        // Deserializes the batch on the host. Then, transforms it to rows and performs row-wise
-        // projection. We should NOT run any device operation on the driver node.
-        val result = withResource(serBatch.value.hostBatches) { hostBatches =>
-          hostBatches.flatMap { cb =>
-            cb.rowIterator().asScala.map { row =>
-              val broadcastRow = broadcastModeProject.map(_(row)).getOrElse(row)
-              rowProject(broadcastRow).copy().asInstanceOf[InternalRow]
-            }
-          }
-        }
-
-        gpuLongMetric("dataSize") += serBatch.value.dataSize
-        gpuLongMetric(COLLECT_TIME) += System.nanoTime() - beforeCollect
 
         result
       }
     }(GpuSubqueryBroadcastExec.executionContext)
+  }
+
+  private def projectSerializedBatchToRows(
+      serBatch: SerializeConcatHostBuffersDeserializeBatch): Array[InternalRow] = {
+    val beforeCollect = System.nanoTime()
+
+    // Creates projection to extract target field from Row, as what Spark does.
+    // Note that unlike Spark, the GPU broadcast data has not applied the key expressions from
+    // the HashedRelation, so that is applied here if necessary to ensure the proper values
+    // are being extracted. The CPU already has the key projections applied in the broadcast
+    // data and thus does not have similar logic here.
+    val broadcastModeProject = modeKeys.map { keyExprs =>
+      val keyExpr = keyExprs(index)
+      UnsafeProjection.create(keyExpr)
+    }
+
+    // Use the single output of the broadcast mode projection if it exists
+    val rowProjectIndex = if (broadcastModeProject.isDefined) 0 else index
+    val rowProject = UnsafeProjection.create(
+      BoundReference(rowProjectIndex, buildKeys(index).dataType, buildKeys(index).nullable))
+
+    // Deserializes the batch on the host. Then, transforms it to rows and performs row-wise
+    // projection. We should NOT run any device operation on the driver node.
+    val result = withResource(serBatch.hostBatches) { hostBatches =>
+      hostBatches.flatMap { cb =>
+        cb.rowIterator().asScala.map { row =>
+          val broadcastRow = broadcastModeProject.map(_(row)).getOrElse(row)
+          rowProject(broadcastRow).copy().asInstanceOf[InternalRow]
+        }
+      }
+    }
+
+    gpuLongMetric("dataSize") += serBatch.dataSize
+    gpuLongMetric(COLLECT_TIME) += System.nanoTime() - beforeCollect
+
+    result
   }
 
   protected override def doExecute(): RDD[InternalRow] = {

